Introduce braket_client (#30)
qiboteam/qibo-cloud-backends: doc/source/tutorials.rst

Extracting Amazon Braket device parameters
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Amazon Braket devices can be found here: https://aws.amazon.com/braket/quantum-computers/. Using the `IQM Garnet device <https://aws.amazon.com/braket/quantum-computers/iqm/>`_ as an example, we demonstrate how to extract useful device information such as the qubit topology and native gates. The qubit connectivity on the IQM Garnet device can be visualised using `NetworkX <https://networkx.org/>`_.

.. code-block:: python

   import networkx as nx

   device = "arn:aws:braket:eu-north-1::device/qpu/iqm/Garnet"
   connectivity_graph = AwsDevice(device).properties.paradigm.connectivity.connectivityGraph
   native_gates = AwsDevice(device).properties.paradigm.nativeGateSet
   print(native_gates)

   G = nx.Graph()
   for node, neighbors in connectivity_graph.items():
      for neighbor in neighbors:
         G.add_edge(node, neighbor)

   nx.draw(G, pos=nx.spring_layout(G), with_labels=True, node_color='lightblue', node_size=500, font_size=10, font_weight='bold', edge_color='gray')

Submit the circuit
~~~~~~~~~~~~~~~~~~

> Let us run a circuit with `verbatim_circuit=True` on an Amazon Braket device, using IQM Garnet as an example. Here, when `verbatim_circuit=True`, the circuit is submitted as is onto the Amazon Braket device. The device expects to receive a circuit that is written in native gates with qubits in the range of the device. For IQM Garnet, the native gates are `CZ` and `PRX` gates. IQM Garnet has qubits indexed from 1 to 20.

Reviewer: ```suggestion
```
this is repeated almost identical in the following line

Author: Oops. Thanks for spotting that!

Therefore, before submitting a Qibo circuit in verbatim mode, it is recommended to extract the Amazon Braket device's information. We will demonstrate this below.

Extracting Amazon Braket device parameters
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

More information on IQM Garnet device can be found here: https://aws.amazon.com/braket/quantum-computers/iqm/.

To use BraketClientBackend, we import these packages first.

.. code-block:: python

   from braket.aws import AwsDevice
   from braket.devices import Devices, LocalSimulator
   from qibo_cloud_backend.braket_client import BraketClientBackend

The qubit connectivity on the IQM Garnet device can be visualised using networkx. We will also extract the device's native gates.

.. code-block:: python

   import networkx as nx

   device = "arn:aws:braket:eu-north-1::device/qpu/iqm/Garnet"
   connectivity_graph = AwsDevice(device).properties.paradigm.connectivity.connectivityGraph
   native_gates = AwsDevice(device).properties.paradigm.nativeGateSet
   print(native_gates)

   G = nx.Graph()
   for node, neighbors in connectivity_graph.items():
      for neighbor in neighbors:
         G.add_edge(node, neighbor)

>    nx.draw(G, pos=nx.spring_layout(G), with_labels=True, node_color='lightblue', node_size=500, font_size=10, font_weight='bold', edge_color='gray')

Author: I feel that it would be useful to leave it here as it's a tutorial about using BraketClientBackend. I'm afraid that links to the API reference might have too much information for the user to extract quickly. I have shortened this segment to the following:

```
Extracting Amazon Braket device parameters
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Amazon Braket devices can be found here: https://aws.amazon.com/braket/quantum-computers/.
Using the `IQM Garnet device <https://aws.amazon.com/braket/quantum-computers/iqm/>`_ as an example,
we demonstrate how to extract useful device information such as the qubit topology and native gates.
The qubit connectivity on the IQM Garnet device can be visualised using `NetworkX <https://networkx.org/>`_.

.. code-block:: python

   import networkx as nx

   device = "arn:aws:braket:eu-north-1::device/qpu/iqm/Garnet"
   connectivity_graph = AwsDevice(device).properties.paradigm.connectivity.connectivityGraph
   native_gates = AwsDevice(device).properties.paradigm.nativeGateSet
   print(native_gates)

   G = nx.Graph()
   for node, neighbors in connectivity_graph.items():
      for neighbor in neighbors:
         G.add_edge(node, neighbor)

   nx.draw(G, pos=nx.spring_layout(G), with_labels=True, node_color='lightblue', 
           node_size=500, font_size=10, font_weight='bold', edge_color='gray')
```

Let me know what you think!

>    nx.draw(G, pos=nx.spring_layout(G), with_labels=True, node_color='lightblue', node_size=500, font_size=10, font_weight='bold', edge_color='gray')

Reviewer: Nice

Submit the circuit
~~~~~~~~~~~~~~~~~~

Let us run a circuit with `verbatim_circuit=True` on an Amazon Braket device, using IQM Garnet as an example. Here, when `verbatim_circuit=True`, the circuit is submitted as is onto the Amazon Braket device. The device expects to receive a circuit that is written in native gates with qubits in the range of the device. For IQM Garnet, the native gates are `CZ` and `PRX` gates. IQM Garnet has qubits indexed from 1 to 20.

.. code-block:: python

   from qibo import gates, Circuit as QiboCircuit
   import numpy as np

   c = QiboCircuit(5)
   c.add(gates.PRX(1, 0.5*np.pi, 1.5*np.pi))

>    c.add(gates.PRX(4, 0.142857142857143*np.pi, 0))

Author: Yes that's right, these are the optimal parameters for this specific graph called the Diamond graph:
![dioamond_graph](https://github.com/user-attachments/assets/80caf72e-df3f-4095-8c21-f8bdc27c9b57)

   c.add(gates.CZ(4, 1))
   c.add(gates.PRX(1, 0.5*np.pi, 0.5*np.pi))
   c.add(gates.M(1))
   c.add(gates.M(4))

   print(c.draw())

We should get this circuit:

.. code-block:: python

   q0: -------------
   q1: -prx-Z-prx-M-
   q2: -----|-------
   q3: -----|-------
   q4: -prx-o-M-----

> Since IQM Garnet has qubits indexed from 1 to 20, we will intentionally leave qubit `q0` empty without any gates. An error will be raised if there are gates on any qubits not in the range from 1 to 20.

Author: Oh. I used qubits `q1` and `q4` because they are coupled (see image of the chip's connectivity [here](https://www.iqmacademy.com/qpu/garnet/). I will just paste the image to save you one click: :-D
![Screenshot 2024-09-23 at 6 04 34 PM](https://github.com/user-attachments/assets/2d545f1d-c2e4-4ad2-94a6-68870aaa4051)

If we submit this circuit with `verbatim_circuit = True`
``` python
   q0: -prx-Z-prx-M-
   q1: -prx-o-M-----
```
there will be an error message because `q0` and `q1` are not connected. But `q1` and `q2` will work. As for using `q1` and `q4`, I've found that the qubit indices need to correspond to the topology, and that creating a circuit with 5 qubits works. So if we were to use the coupled qubits `q19` and `q20`, assuming that 19 and 20 have the lowest error rates etc.., then we'll need to create a 21 qubit circuit.

That said, would you know if it is possible to create a qibo circuit like this, without having to create a 21 qubit circuit?
``` python
   q19: -prx-Z-prx-M-
   q20: -prx-o-M-----
```
If it's possible, then I think running it in verbatim mode on the Amazon Braket device might be doable!

> Since IQM Garnet has qubits indexed from 1 to 20, we will intentionally leave qubit `q0` empty without any gates. An error will be raised if there are gates on any qubits not in the range from 1 to 20.

Reviewer: Ah ok I see, so here you are running with `verbatim=True`. Then maybe, try to include the plot of the topology in the tutorial and explicitely explain that since `verbatim=True` you have to create the circuit this way

   print(counts)

For completeness, one can also use the LocalSimulator to execute circuit `c` as follows.

.. code-block:: python

   device = LocalSimulator("default")
   AWS = BraketClientBackend(device = device, verbatim_circuit=True)

   counts = AWS.execute_circuit(c, nshots=1000).frequencies()
   print(counts)

Example: Using Zero Noise Extrapolation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In this example, we illustrate the use of Zero Noise Extrapolation (ZNE) to improve the results of a Quantum Approximate Optimization Algorithm (QAOA) circuit. The circuit solves a trivial MaxCut problem with a single QAOA layer. The backend for ZNE is set to an Amazon Braket device.

Here, we make several assumptions:
1. The user is able to transpile any Qibo circuit with IQM Garnet's qubit topology.
2. The optimal angles for the single QAOA layer are found and will be used.

With these assumptions met, we provide an example circuit `c` below is written in IQM Garnet's native gates with specific qubits. The parameters for the `PRX` gates are optimal. We select `verbatim_circuit=True` as we do not want the device to transpile the circuit.

.. code-block:: python

>    c = QiboCircuit(10):

Author: I'm not quite sure how we can do this. The QAOA for the Diamond graph gives this circuit:
``` python
q0: ─H─o────o─o────o─o────o───────────────RX─M─
q1: ─H─X─RZ─X─|────|─|────|─o────o────────RX─M─
q2: ─H────────X─RZ─X─|────|─X─RZ─X─o────o─RX─M─
q3: ─H───────────────X─RZ─X────────X─RZ─X─RX─M─
```

A couple of months ago when we did this, we used the Qiskit's transpiler to get the transpiled circuit. IQM's `coupling_map` and the `['r', 'cz']` native gates were given as inputs to Qiskit's transpiler and it returned this huge circuit. (Recall `r` = `prx` gate)

>    c = QiboCircuit(10):

Reviewer: Mmmh yeah I figured that. The un-transpiled circuit is actually pretty simple to shorten as you have the repeated `X RZ X` pattern, however after transpilation this probably gets messed up and it's not easy to find repeated patterns anymore. Well I guess we have to deal with that.

>    c = QiboCircuit(10):

Author: Got it!

   c.add(gates.PRX(5, -1.081592653589793, 0))
   c.add(gates.PRX(5, np.pi, -np.pi))
   c.add(gates.CZ(4, 5))
   c.add(gates.PRX(4, 2.850796326794897, 0))
   c.add(gates.PRX(5, -np.pi, np.pi/2))
   c.add(gates.PRX(5, np.pi, -np.pi/2))
   c.add(gates.PRX(9, -np.pi/2, -np.pi))
   c.add(gates.PRX(9, np.pi, -np.pi/4))
   c.add(gates.CZ(4, 9))
   c.add(gates.PRX(4, np.pi/2, 0))
   c.add(gates.PRX(9, np.pi/2, 0))
   c.add(gates.CZ(4, 9))
   c.add(gates.PRX(4, np.pi/2, 0))
   c.add(gates.PRX(9, np.pi/2, 0))
   c.add(gates.CZ(4, 9))
   c.add(gates.PRX(4, np.pi/2, np.pi/2))
   c.add(gates.PRX(4, np.pi, 0))
   c.add(gates.CZ(3, 4))
   c.add(gates.PRX(4, -1.081592653589793, 0))
   c.add(gates.PRX(4, np.pi, -np.pi))
   c.add(gates.CZ(3, 4))
   c.add(gates.PRX(3, 1.28, 0))
   c.add(gates.PRX(4, np.pi/2, np.pi/2))
   c.add(gates.PRX(4, np.pi, 0))
   c.add(gates.CZ(4, 5))
   c.add(gates.PRX(4, -1.081592653589793, 0))
   c.add(gates.PRX(4, np.pi, -np.pi))
   c.add(gates.CZ(4, 5))
   c.add(gates.PRX(4, 1.28, 0))
   c.add(gates.PRX(5, -np.pi/2, -2.850796326794897))
   c.add(gates.PRX(5, np.pi, -0.64)
   c.add(gates.M(9))
   c.add(gates.M(3))
   c.add(gates.M(4))
   c.add(gates.M(5))

> We define the problem Hamiltonian of QAOA for MaxCut, `obs`. The `obs` has to be written according to the qubit mapping applied for circuit `c`.

Reviewer: ```suggestion
We define the problem Hamiltonian of the QAOA for MaxCut, `obs`, adapted to fit the manually transpiled circuit constructed above.
```

.. code-block:: python

   obs = 2.5 - 0.5*Z(3)*Z(9) - 0.5*Z(4)*Z(3) - 0.5*Z(4)*Z(5) - 0.5*Z(4)*Z(9) - 0.5*Z(9)*Z(5)
   obs = SymbolicHamiltonian(obs, nqubits=c.nqubits, backend=NumpyBackend())

Finally, we can run ZNE by setting the backend to the `AWS` to obtain the estimated (extrapolated) result.

.. code-block:: python

>    device = AwsDevice('arn:aws:braket:eu-north-1::device/qpu/iqm/Garnet')
>    AWS = BraketClientBackend(device = device, verbatim_circuit=True)

Reviewer: ```suggestion
```
Again, I would avoid repeating this over and over. I would just define it on top and, as long as it is unchanged, refer to it in the following.

.. code-block:: python

   from qibo import gates, Circuit as QiboCircuit

   c = QiboCircuit(2)
   c.add(gates.H(0))
   c.add(gates.CNOT(0, 1))
   c.add(gates.M(0))
   c.add(gates.M(1))

   print(c.draw())

We should get this circuit:

.. code-block:: python

   q0: -H-o-M-
   q1: ---x-M-

Now, we initialize a Amazon Braket device (e.g. IQM Garnet) and execute circuit `c` on the backend `AWS`.

.. code-block:: python

   device = "arn:aws:braket:eu-north-1::device/qpu/iqm/Garnet"
   AWS = BraketClientBackend(device = device, verbatim_circuit=False)

   counts = AWS.execute_circuit(c, nshots=1000).frequencies()
   print(counts)

Alternatively, one can also use the LocalSimulator to execute circuit `c`. This can be done by leaving `device` empty as it defaults to the LocalSimulator.

.. code-block:: python

   AWS = BraketClientBackend(verbatim_circuit=False)

   counts = AWS.execute_circuit(c, nshots=1000).frequencies()
   print(counts)

To monitor the status of a circuit that is executed, especially on an Amazon Braket device, one can set `monitoring=True`. By default, `monitoring=False`.

.. code-block:: python

   device = "arn:aws:braket:eu-north-1::device/qpu/iqm/Garnet"

>    AWS = BraketClientBackend(device = device, verbatim_circuit=False, monitoring=True)

Reviewer: ```suggestion
   AWS = BraketClientBackend(device = device, verbatim_circuit=False, verbosity=True)
```

Author: I forgot to change the documentation. Will do that!
